File: Longest Palindromic Substring/solution_optimal.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        dp = [[False for _ in range(n)] for _ in range(n)]

        for i in range(n):
            dp[i][i] = True
        
        maxstr = s[0]

        for i in range(n-2, -1, -1):
            for j in range(i+1, n):
                dp[i][j] = s[i] == s[j] and (j-i == 1 or dp[i+1][j-1])

                if dp[i][j] and len(maxstr) < j-i+1:
                    maxstr = s[i:j+1]

        return maxstr

File: Longest Palindromic Substring/solution_optimal.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.length();
        

        vector<vector<bool>> dp(n, vector<bool>(n, false));

        for (int i = 0; i < n; ++i)
            dp[i][i] = true;

        string maxstr = s.substr(0, 1);
        for (int i = n-2; i >= 0; --i)
        {
            for (int j = i+1; j < n; ++j)
            {
                if (s[i] == s[j] && (j-i == 1 || dp[i+1][j-1]))
                {
                    dp[i][j] = true;

                    if (j-i+1 > maxstr.length())
                        maxstr = s.substr(i, j-i+1);
                }
            }
        }

        return maxstr;
    }
};